File: generate_api_validation_workbook.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import ExcelJS from 'exceljs';
import * as yaml from 'js-yaml';

const args = process.argv.slice(2);
const valueAfter = (name) => {
  const index = args.indexOf(name);
  return index >= 0 ? args[index + 1] : undefined;
};
const inputRoot = valueAfter('--input');
const output = valueAfter('--output');
if (!inputRoot || !output) {
  console.error('Usage: node generate_api_validation_workbook.mjs --input <yaml-root> --output <xlsx>');
  process.exit(2);
}

const headers = ['Test Case ID', 'Test Case Title', 'Test Type', 'Test Category', 'HTTP Method', 'Endpoint', 'Pre-conditions', 'Test Steps', 'Input Data', 'Expected Result', 'Priority'];
const headerFill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF17365D' } };
const priorityFill = {
  High: { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFF4CCCC' } },
  Medium: { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFFFF2CC' } },
  Low: { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFD9EAD3' } },
};
const widths = [8, 34, 18, 24, 12, 52, 38, 48, 58, 68, 12];

function text(value) {
  if (value === null || value === undefined) return '';
  if (typeof value === 'object') return JSON.stringify(value, null, 2);
  return String(value);
}
function block(value) {
  if (value === null || value === undefined || value === '' || (typeof value === 'object' && Object.keys(value).length === 0)) return '{}';
  return yaml.dump(value, { noRefs: true, sortKeys: false, lineWidth: -1 }).trim();
}
function priority(value) { return ({ P1: 'High', P2: 'Medium', P3: 'Low' })[text(value).toUpperCase()] || 'Medium'; }
function type(tags, title) {
  const set = new Set((tags || []).map((tag) => text(tag).toLowerCase()));
  const lower = title.toLowerCase();
  if (set.has('edge-case') || set.has('resilience') || lower.includes('timeout') || lower.includes('downstream')) return 'Edge/Resilience';
  if (set.has('boundary')) return 'Boundary';
  if (set.has('negative')) return 'Negative';
  return 'Positive';
}
function category(tags, title) {
  const set = new Set((tags || []).map((tag) => text(tag).toLowerCase()));
  const lower = title.toLowerCase();
  if (set.has('authentication')) return 'Authentication';
  if (set.has('security')) return 'Security';
  if (set.has('authorization') || lower.includes('role')) return 'Authorization';
  if (lower.includes('rate') || lower.includes('thrott')) return 'Rate Limiting';
  if (lower.includes('content-type') || set.has('http')) return 'HTTP Protocol';
  if (set.has('boundary')) return 'Boundary Testing';
  if (set.has('validation')) return 'Field Validation';
  if (lower.includes('duplicate') || lower.includes('timeout') || lower.includes('empty')) return 'Error Handling';
  return set.has('business') ? 'Business Rule' : 'API Contract';
}
function expected(caseData) {
  const response = caseData.response || {};
  const lines = [];
  if (response.successStatusCode !== undefined) lines.push(`HTTP status: ${response.successStatusCode}`);
  if (response.successDescription) lines.push(`Response description: ${response.successDescription}`);
  for (const assertion of caseData.assertions || []) lines.push(`Assertion: ${text(assertion)}`);
  for (const field of ['schema', 'bodySchema', 'expectedFields']) if (response[field]) lines.push(`${field}:\n${block(response[field])}`);
  if (response.errorScenarios) lines.push(`Error scenarios:\n${block(response.errorScenarios)}`);
  if (response.errorStatusCodes) lines.push(`Error status codes:\n${block(response.errorStatusCodes)}`);
  return lines.length ? lines.join('\n') : 'Assumption: The YAML does not define an expected status code or response body.';
}
function row(caseData) {
  const request = caseData.request || {};
  const title = text(caseData.title || caseData.id);
  return [
    text(caseData.id), title, type(caseData.tags, title), category(caseData.tags, title),
    text(request.method || caseData.method), text(request.url || request.endpoint || caseData.endpoint),
    block(caseData.preconditions || []),
    ['1. Prepare the pre-conditions and test data described for this scenario.', `2. Send the ${request.method || 'specified'} request to the endpoint with the listed headers, parameters, and body.`, '3. Capture the HTTP status code and response body.', '4. Compare the response with every expected result and assertion.'].join('\n'),
    block({ headers: request.headers || {}, pathParams: request.pathParams || {}, queryParams: request.queryParams || {}, body: request.body || {} }),
    expected(caseData), priority(caseData.priority),
  ];
}
function allYamlFiles(directory) {
  const result = [];
  for (const entry of fs.readdirSync(directory, { withFileTypes: true })) {
    const full = path.join(directory, entry.name);
    if (entry.isDirectory()) result.push(...allYamlFiles(full));
    else if (entry.isFile() && entry.name.endsWith('.yaml') && entry.name !== '_index.yaml') result.push(full);
  }
  return result.sort();
}
function loadInputs() {
  const grouped = new Map();
  const inventory = [];
  for (const entry of fs.readdirSync(inputRoot, { withFileTypes: true }).sort((a, b) => a.name.localeCompare(b.name))) {
    if (!entry.isDirectory() || entry.name.toLowerCase() === 'test cases') continue;
    const cases = [];
    for (const yamlFile of allYamlFiles(path.join(inputRoot, entry.name))) {
      const data = yaml.load(fs.readFileSync(yamlFile, 'utf8')) || {};
      if (!data || typeof data !== 'object' || !data.id) continue;
      cases.push({ values: row(data), file: yamlFile, data });
    }
    if (cases.length) {
      const first = cases[0].data;
      const request = first.request || {};
      const refs = first.references || {};
      inventory.push({ api: entry.name, method: text(request.method || first.method), endpoint: text(request.url || request.endpoint || first.endpoint), sourceWorksheet: text(refs.sourceWorksheet) || 'Assumption: not present in YAML', apiSpecFile: text(refs.apiSpecFile) || 'Assumption: not present in YAML' });
      grouped.set(entry.name, cases);
    }
  }
  return { grouped, inventory };
}
function formatCaseSheet(ws, cases) {
  ws.views = [{ state: 'frozen', ySplit: 1 }];
  ws.autoFilter = { from: 'A1', to: `K${Math.max(1, cases.length + 1)}` };
  headers.forEach((header, index) => {
    const cell = ws.getCell(1, index + 1);
    cell.value = header; cell.fill = headerFill; cell.font = { bold: true, color: { argb: 'FFFFFFFF' } }; cell.alignment = { horizontal: 'center', vertical: 'middle', wrapText: true };
  });
  cases.forEach(({ values, file, data }, rowIndex) => {
    const rowNumber = rowIndex + 2;
    values.forEach((value, columnIndex) => { ws.getCell(rowNumber, columnIndex + 1).value = value; ws.getCell(rowNumber, columnIndex + 1).alignment = { vertical: 'top', wrapText: true }; });
    ws.getCell(rowNumber, 1).note = `Source YAML: ${file}\nSource worksheet: ${text((data.references || {}).sourceWorksheet)}\nAPI spec file: ${text((data.references || {}).apiSpecFile)}`;
    ws.getCell(rowNumber, 11).fill = priorityFill[values[10]];
    ws.getRow(rowNumber).height = 92;
  });
  widths.forEach((width, index) => { ws.getColumn(index + 1).width = width; });
  ws.getRow(1).height = 34;
}
function addHeader(row, values) {
  values.forEach((value, index) => { const cell = row.getCell(index + 1); cell.value = value; cell.fill = headerFill; cell.font = { bold: true, color: { argb: 'FFFFFFFF' } }; });
}
async function main() {
  const { grouped, inventory } = loadInputs();
  if (!grouped.size) throw new Error(`No executable YAML scenarios found under ${inputRoot}`);
  const allCases = [...grouped.values()].flat();
  const counts = Object.fromEntries(['Positive', 'Negative', 'Boundary', 'Edge/Resilience'].map((key) => [key, allCases.filter((item) => item.values[2] === key).length]));
  const categories = [...new Set(allCases.map((item) => item.values[3]))].sort().map((key) => [key, allCases.filter((item) => item.values[3] === key).length]);
  const workbook = new ExcelJS.Workbook();
  const summary = workbook.addWorksheet('Summary');
  summary.getCell('A1').value = 'ECLIPSE App Dashboard API Validation Test Cases'; summary.getCell('A1').font = { bold: true, size: 16, color: { argb: 'FF17365D' } };
  summary.getCell('A3').value = 'Generated date'; summary.getCell('B3').value = new Date().toISOString();
  [['Total APIs covered', grouped.size], ['Total test cases', allCases.length], ...Object.entries(counts)].forEach(([label, value], index) => { summary.getCell(index + 5, 1).value = label; summary.getCell(index + 5, 1).font = { bold: true }; summary.getCell(index + 5, 2).value = value; summary.getCell(index + 5, 2).font = { bold: true }; });
  const categoryStart = 12; summary.getCell(categoryStart, 1).value = 'Test category breakdown'; summary.getCell(categoryStart, 1).font = { bold: true, size: 12 }; addHeader(summary.getRow(categoryStart + 1), ['Category', 'Count']);
  categories.forEach(([key, count], index) => { summary.getCell(categoryStart + 2 + index, 1).value = key; summary.getCell(categoryStart + 2 + index, 2).value = count; });
  const inventoryStart = categoryStart + categories.length + 5; summary.getCell(inventoryStart, 1).value = 'API inventory'; summary.getCell(inventoryStart, 1).font = { bold: true, size: 12 }; addHeader(summary.getRow(inventoryStart + 1), ['API folder', 'HTTP Method', 'Endpoint', 'Source worksheet', 'API spec file']);
  inventory.forEach((item, index) => { ['api', 'method', 'endpoint', 'sourceWorksheet', 'apiSpecFile'].forEach((key, column) => { summary.getCell(inventoryStart + 2 + index, column + 1).value = item[key]; summary.getCell(inventoryStart + 2 + index, column + 1).alignment = { vertical: 'top', wrapText: true }; }); });
  [28, 18, 52, 34, 42].forEach((width, index) => { summary.getColumn(index + 1).width = width; });
  summary.views = [{ state: 'frozen', ySplit: inventoryStart + 1 }];
  const validation = workbook.addWorksheet('Validation Areas'); validation.getCell('A1').value = 'KEY VALIDATION AREAS COVERED'; validation.getCell('A1').font = { bold: true, size: 14, color: { argb: 'FF17365D' } }; addHeader(validation.getRow(3), ['Area', 'Coverage']);
  const areas = ['Field validation: mandatory, optional, null, and omitted fields.', 'Authentication: missing and invalid credentials.', 'Authorization: role and permission checks where represented.', 'Business rules: assertions, duplicates, and contract behavior.', 'Data types and formats: schema, enum, date, numeric, and format assertions.', 'Boundary conditions: minimum, maximum, null, omitted, and invalid boundaries.', 'Error handling: malformed or empty requests and downstream failures.', 'Security: security-tagged authentication and invalid-input scenarios.', 'HTTP protocol: method, endpoint, content type, and status checks.', 'Rate limiting/performance: included when represented; otherwise not defined.'];
  areas.forEach((area, index) => { const [name, ...rest] = area.split(':'); validation.getCell(index + 4, 1).value = name; validation.getCell(index + 4, 2).value = rest.join(':'); validation.getCell(index + 4, 2).alignment = { wrapText: true, vertical: 'top' }; validation.getRow(index + 4).height = 34; }); validation.getColumn(1).width = 30; validation.getColumn(2).width = 110;
  for (const [api, cases] of grouped) formatCaseSheet(workbook.addWorksheet(api.slice(0, 31)), cases);
  formatCaseSheet(workbook.addWorksheet('All Test Cases'), allCases);
  fs.mkdirSync(path.dirname(output), { recursive: true }); await workbook.xlsx.writeFile(output);
  console.log(`Generated ${output} with ${grouped.size} APIs and ${allCases.length} test cases`);
}
main().catch((error) => { console.error(error.stack || error); process.exit(1); });

File: generate_api_validation_workbook.py
#!/usr/bin/env python3
"""Generate a formatted API validation workbook from generated BDD YAML files."""

from __future__ import annotations

import argparse
import json
from collections import Counter, defaultdict
from datetime import datetime
from pathlib import Path
from typing import Any

import yaml
from openpyxl import Workbook
from openpyxl.comments import Comment
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.utils import get_column_letter


HEADERS = [
    "Test Case ID", "Test Case Title", "Test Type", "Test Category",
    "HTTP Method", "Endpoint", "Pre-conditions", "Test Steps",
    "Input Data", "Expected Result", "Priority",
]
API_SHEET_COLUMNS = [8, 34, 18, 24, 12, 52, 38, 48, 58, 68, 12]
HEADER_FILL = PatternFill("solid", fgColor="17365D")
TITLE_FILL = PatternFill("solid", fgColor="D9EAF7")
PRIORITY_FILLS = {
    "High": PatternFill("solid", fgColor="F4CCCC"),
    "Medium": PatternFill("solid", fgColor="FFF2CC"),
    "Low": PatternFill("solid", fgColor="D9EAD3"),
}


def as_text(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, str):
        return value
    if isinstance(value, (dict, list)):
        return json.dumps(value, indent=2, ensure_ascii=True, default=str)
    return str(value)


def yaml_block(value: Any) -> str:
    if value in (None, {}, [], ""):
        return "{}"
    return yaml.safe_dump(value, sort_keys=False, allow_unicode=False, default_flow_style=False).strip()


def priority(value: Any) -> str:
    return {"P1": "High", "P2": "Medium", "P3": "Low"}.get(str(value).upper(), "Medium")


def test_type(tags: list[str], title: str) -> str:
    values = {str(tag).lower() for tag in tags}
    if "edge-case" in values or "resilience" in values or any(x in title.lower() for x in ("timeout", "downstream")):
        return "Edge/Resilience"
    if "boundary" in values:
        return "Boundary"
    if "negative" in values:
        return "Negative"
    return "Positive"


def category(tags: list[str], title: str) -> str:
    values = {str(tag).lower() for tag in tags}
    title_lower = title.lower()
    if values & {"authentication", "security"}:
        return "Authentication" if "authentication" in values else "Security"
    if "authorization" in values or "role" in title_lower:
        return "Authorization"
    if "rate" in title_lower or "thrott" in title_lower:
        return "Rate Limiting"
    if "content-type" in title_lower or "http" in values:
        return "HTTP Protocol"
    if values & {"boundary", "validation"}:
        return "Boundary Testing" if "boundary" in values else "Field Validation"
    if "duplicate" in title_lower or "timeout" in title_lower or "empty" in title_lower:
        return "Error Handling"
    return "Business Rule" if "business" in values else "API Contract"


def steps_for(case: dict[str, Any]) -> str:
    request = case.get("request") or {}
    return "\n".join([
        "1. Prepare the pre-conditions and test data described for this scenario.",
        f"2. Send the {request.get('method', 'specified')} request to the endpoint with the listed headers, parameters, and body.",
        "3. Capture the HTTP status code and response body.",
        "4. Compare the response with every expected result and assertion.",
    ])


def expected_result(case: dict[str, Any]) -> str:
    response = case.get("response") or {}
    lines: list[str] = []
    if response.get("successStatusCode") is not None:
        lines.append(f"HTTP status: {response['successStatusCode']}")
    if response.get("successDescription"):
        lines.append(f"Response description: {response['successDescription']}")
    for assertion in case.get("assertions") or []:
        lines.append(f"Assertion: {as_text(assertion)}")
    for field in ("schema", "bodySchema", "expectedFields"):
        if response.get(field):
            lines.append(f"{field}:\n{yaml_block(response[field])}")
    if response.get("errorScenarios"):
        lines.append(f"Error scenarios:\n{yaml_block(response['errorScenarios'])}")
    if response.get("errorStatusCodes"):
        lines.append(f"Error status codes:\n{yaml_block(response['errorStatusCodes'])}")
    if not lines:
        lines.append("Assumption: The YAML does not define an expected status code or response body.")
    return "\n".join(lines)


def make_row(case: dict[str, Any], api_folder: str, source_yaml: Path) -> list[str]:
    request = case.get("request") or {}
    tags = case.get("tags") or []
    title = as_text(case.get("title") or case.get("id"))
    input_data = {
        "headers": request.get("headers") or {},
        "pathParams": request.get("pathParams") or {},
        "queryParams": request.get("queryParams") or {},
        "body": request.get("body") or {},
    }
    return [
        as_text(case.get("id")), title, test_type(tags, title), category(tags, title),
        as_text(request.get("method") or case.get("method")),
        as_text(request.get("url") or request.get("endpoint") or case.get("endpoint")),
        yaml_block(case.get("preconditions") or []), steps_for(case), yaml_block(input_data),
        expected_result(case), priority(case.get("priority")),
    ]


def load_cases(input_root: Path) -> tuple[dict[str, list[tuple[list[str], Path, dict[str, Any]]]], list[dict[str, str]]]:
    grouped: dict[str, list[tuple[list[str], Path, dict[str, Any]]]] = defaultdict(list)
    inventory: list[dict[str, str]] = []
    for api_dir in sorted(path for path in input_root.iterdir() if path.is_dir() and path.name.lower() != "test cases"):
        cases = []
        for yaml_path in sorted(api_dir.rglob("*.yaml")):
            if yaml_path.name == "_index.yaml":
                continue
            with yaml_path.open("r", encoding="utf-8") as stream:
                case = yaml.safe_load(stream) or {}
            if not isinstance(case, dict) or not case.get("id"):
                continue
            cases.append((make_row(case, api_dir.name, yaml_path), yaml_path, case))
        if cases:
            first = cases[0][2]
            request = first.get("request") or {}
            references = first.get("references") or {}
            inventory.append({
                "api": api_dir.name,
                "method": as_text(request.get("method") or first.get("method")),
                "endpoint": as_text(request.get("url") or request.get("endpoint") or first.get("endpoint")),
                "sourceWorksheet": as_text(references.get("sourceWorksheet")) or "Assumption: not present in YAML",
                "apiSpecFile": as_text(references.get("apiSpecFile")) or "Assumption: not present in YAML",
            })
            grouped[api_dir.name] = cases
    return grouped, inventory


def style_case_sheet(ws, rows: list[tuple[list[str], Path, dict[str, Any]]]) -> None:
    ws.freeze_panes = "A2"
    ws.auto_filter.ref = f"A1:K{max(1, len(rows) + 1)}"
    for index, header in enumerate(HEADERS, 1):
        cell = ws.cell(1, index, header)
        cell.fill = HEADER_FILL
        cell.font = Font(bold=True, color="FFFFFF")
        cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)
    for row_number, (values, yaml_path, case) in enumerate(rows, 2):
        for column, value in enumerate(values, 1):
            cell = ws.cell(row_number, column, value)
            cell.alignment = Alignment(vertical="top", wrap_text=True)
        ws.cell(row_number, 1).comment = Comment(
            "Source YAML: " + str(yaml_path) + "\nSource worksheet: " +
            as_text((case.get("references") or {}).get("sourceWorksheet")) +
            "\nAPI spec file: " + as_text((case.get("references") or {}).get("apiSpecFile")),
            "QA/API Test Generator",
        )
        ws.cell(row_number, 11).fill = PRIORITY_FILLS[values[10]]
        ws.row_dimensions[row_number].height = 92
    for column, width in enumerate(API_SHEET_COLUMNS, 1):
        ws.column_dimensions[get_column_letter(column)].width = width
    ws.row_dimensions[1].height = 34


def build_workbook(grouped: dict[str, list[tuple[list[str], Path, dict[str, Any]]]], inventory: list[dict[str, str]], output: Path) -> None:
    all_rows = [item for api_rows in grouped.values() for item in api_rows]
    counts = Counter(item[0][2] for item in all_rows)
    categories = Counter(item[0][3] for item in all_rows)
    wb = Workbook()
    summary = wb.active
    summary.title = "Summary"
    summary["A1"] = "ECLIPSE App Dashboard API Validation Test Cases"
    summary["A1"].font = Font(bold=True, size=16, color="17365D")
    summary["A3"] = "Generated date"
    summary["B3"] = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    metrics = [("Total APIs covered", len(grouped)), ("Total test cases", len(all_rows)),
               ("Positive", counts["Positive"]), ("Negative", counts["Negative"]),
               ("Boundary", counts["Boundary"]), ("Edge/Resilience", counts["Edge/Resilience"])]
    for row, (label, value) in enumerate(metrics, 5):
        summary.cell(row, 1, label).font = Font(bold=True)
        summary.cell(row, 2, value).font = Font(bold=True)
    summary["A13"] = "Test category breakdown"
    summary["A13"].font = Font(bold=True, size=12)
    summary.append([])
    summary.cell(14, 1, "Category").fill = HEADER_FILL
    summary.cell(14, 2, "Count").fill = HEADER_FILL
    for cell in summary[14][:2]:
        cell.font = Font(bold=True, color="FFFFFF")
    for row, (label, count) in enumerate(sorted(categories.items()), 15):
        summary.cell(row, 1, label)
        summary.cell(row, 2, count)
    inventory_row = 15 + len(categories) + 2
    summary.cell(inventory_row, 1, "API inventory").font = Font(bold=True, size=12)
    columns = ["API folder", "HTTP Method", "Endpoint", "Source worksheet", "API spec file"]
    for column, label in enumerate(columns, 1):
        cell = summary.cell(inventory_row + 1, column, label)
        cell.fill = HEADER_FILL
        cell.font = Font(bold=True, color="FFFFFF")
    for row, item in enumerate(inventory, inventory_row + 2):
        for column, key in enumerate(("api", "method", "endpoint", "sourceWorksheet", "apiSpecFile"), 1):
            summary.cell(row, column, item[key]).alignment = Alignment(vertical="top", wrap_text=True)
    summary.column_dimensions["A"].width = 28
    for column in "BCDE":
        summary.column_dimensions[column].width = 42
    summary.freeze_panes = "A15"

    validation = wb.create_sheet("Validation Areas")
    validation["A1"] = "KEY VALIDATION AREAS COVERED"
    validation["A1"].font = Font(bold=True, size=14, color="17365D")
    areas = [
        "Field validation: mandatory, optional, null, and omitted fields from YAML scenarios.",
        "Authentication: missing and invalid authentication credentials.",
        "Authorization: role and permission checks where represented by scenario tags or titles.",
        "Business rules: scenario assertions, duplicate submissions, and contract behavior.",
        "Data types and formats: schema, enum, date, numeric, and format assertions from YAML.",
        "Boundary conditions: minimum, maximum, null, omitted, and invalid boundary scenarios.",
        "Error handling: malformed or empty requests, downstream failures, and expected errors.",
        "Security: security-tagged authentication and invalid-input scenarios.",
        "HTTP protocol: method, endpoint, content type, and expected status code checks.",
        "Rate limiting/performance: included when represented in the generated YAML; otherwise marked as not defined.",
    ]
    validation["A3"] = "Area"
    validation["B3"] = "Coverage"
    for cell in validation[3][:2]:
        cell.fill = HEADER_FILL
        cell.font = Font(bold=True, color="FFFFFF")
    for row, area in enumerate(areas, 4):
        validation.cell(row, 1, area.split(":", 1)[0])
        validation.cell(row, 2, area).alignment = Alignment(wrap_text=True, vertical="top")
        validation.row_dimensions[row].height = 34
    validation.column_dimensions["A"].width = 30
    validation.column_dimensions["B"].width = 110

    for api_name, rows in grouped.items():
        ws = wb.create_sheet(api_name[:31])
        style_case_sheet(ws, rows)
    all_sheet = wb.create_sheet("All Test Cases")
    style_case_sheet(all_sheet, all_rows)
    output.parent.mkdir(parents=True, exist_ok=True)
    wb.save(output)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", type=Path, required=True, help="Generated API YAML root")
    parser.add_argument("--output", type=Path, required=True, help="Output .xlsx path")
    args = parser.parse_args()
    grouped, inventory = load_cases(args.input)
    if not grouped:
        raise SystemExit(f"No executable YAML scenarios found under {args.input}")
    build_workbook(grouped, inventory, args.output)
    print(f"Generated {args.output} with {len(grouped)} APIs and {sum(map(len, grouped.values()))} test cases")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())